src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs

```diff
@@ -45,76 +42,96 @@ private static async ValueTask<TValue> ReadAsync<TValue>(Stream utf8Json, Type r
                 state.Current.JsonPropertyInfo = classInfo.GetPolicyProperty();
             }
 
-            var readerState = new JsonReaderState(options: options.ReaderOptions);
-
-            int bytesRemaining = 0;
-            int bytesRead;
+            var readerState = new JsonReaderState(options.ReaderOptions);
 
             // todo: switch to ArrayBuffer implementation to handle and simplify the allocs?
             byte[] buffer = ArrayPool<byte>.Shared.Rent(options.EffectiveBufferSize);
-            int bufferSize = buffer.Length;
-            int deserializeBufferSize;
-            bool isFinalBlock;
+            int bytesInBuffer = 0;
+            long totalBytesRead = 0;
+            int clearMax = 0;
 
             try
             {
-                do
+                while (true)
                 {
-                    int bytesToRead = bufferSize - bytesRemaining;
-                    bytesRead = await utf8Json.ReadAsync(buffer, bytesRemaining, bytesToRead, cancellationToken).ConfigureAwait(false);
+                    // Read from the stream until either our buffer is filled or we hit EOF.
+                    // Calling ReadCore is relatively expensive, so we minimize the number of times
+                    // we need to call it.
+                    bool isFinalBlock = false;
+                    while (true)
+                    {
+                        int bytesRead = await utf8Json.ReadAsync(buffer.AsMemory(bytesInBuffer), cancellationToken).ConfigureAwait(false);
+
+                        if (bytesRead == 0)
+                        {
+                            isFinalBlock = true;
+                            break;
+                        }
 
-                    deserializeBufferSize = bytesRemaining + bytesRead;
-                    isFinalBlock = (bytesRead == 0);
+                        totalBytesRead += bytesRead;
+                        bytesInBuffer += bytesRead;
 
+                        if (bytesInBuffer > clearMax)
+                        {
+                            clearMax = bytesRead;
+                        }
+
+                        if (bytesInBuffer == buffer.Length)
+                        {
+                            break;
+                        }
+                    }
+
+                    // Process the data available
                     ReadCore(
                         ref readerState,
                         isFinalBlock,
-                        buffer,
-                        deserializeBufferSize,
+                        new Span<byte>(buffer, 0, bytesInBuffer),
                         options,
                         ref state);
 
                     int bytesConsumed = (int)readerState.BytesConsumed;
-                    bytesRemaining = deserializeBufferSize - bytesConsumed;
+                    bytesInBuffer -= bytesConsumed;
```

Comment:
unrelated change, but: `Debug.Assert` that `readerState.BytesConsumed` is `<= int.MaxValue` before casting to `(int)`.

Reply:
It would actually be <= bytesInBuffer, right?